File: ios-sdk/Sources/GazePointSDK/KalmanFilter.swift
import Foundation
import CoreGraphics

/// Kalman Filter for smoothing gaze point tracking
public class KalmanFilter {
    
    private var estimateX: CGFloat = 0
    private var estimateY: CGFloat = 0
    private var errorCovarianceX: CGFloat = 1
    private var errorCovarianceY: CGFloat = 1
    
    private let processNoise: CGFloat = 0.01
    private let measurementNoise: CGFloat = 0.1
    
    public init() {}
    
    /// Update the filter with a new measurement
    public func update(measurement: CGPoint) -> CGPoint {
        // Prediction step (assume no change in position)
        let predictedErrorCovX = errorCovarianceX + processNoise
        let predictedErrorCovY = errorCovarianceY + processNoise
        
        // Update step
        let kalmanGainX = predictedErrorCovX / (predictedErrorCovX + measurementNoise)
        let kalmanGainY = predictedErrorCovY / (predictedErrorCovY + measurementNoise)
        
        estimateX += kalmanGainX * (measurement.x - estimateX)
        estimateY += kalmanGainY * (measurement.y - estimateY)
        
        errorCovarianceX = (1 - kalmanGainX) * predictedErrorCovX
        errorCovarianceY = (1 - kalmanGainY) * predictedErrorCovY
        
        return CGPoint(x: estimateX, y: estimateY)
    }
    
    /// Reset the filter to initial state
    public func reset() {
        estimateX = 0
        estimateY = 0
        errorCovarianceX = 1
        errorCovarianceY = 1
    }
}

File: web/src/utils/KalmanFilter.ts
/**
 * Kalman Filter for smoothing gaze point tracking
 */

export class KalmanFilter {
  private processNoise: number;
  private measurementNoise: number;
  private estimate: number = 0;
  private errorCovariance: number = 1;

  constructor(processNoise: number = 0.01, measurementNoise: number = 0.1) {
    this.processNoise = processNoise;
    this.measurementNoise = measurementNoise;
  }

  public filter(measurement: number): number {
    // Prediction
    const predictedEstimate = this.estimate;
    const predictedErrorCovariance = this.errorCovariance + this.processNoise;

    // Update
    const kalmanGain =
      predictedErrorCovariance /
      (predictedErrorCovariance + this.measurementNoise);
    
    this.estimate = predictedEstimate + kalmanGain * (measurement - predictedEstimate);
    this.errorCovariance = (1 - kalmanGain) * predictedErrorCovariance;

    return this.estimate;
  }

  public reset(): void {
    this.estimate = 0;
    this.errorCovariance = 1;
  }
}
